docs/ldml/tr35-keyboards.md

```diff
@@ -478,6 +478,10 @@ _Attribute:_ `normalization`
 
 > The `normalization` attribute describes the intended normalization form of the keyboard layout output. The valid values are `NFC`, `NFD` or `other`.
 > An example use case is aiding a user to choose among the two same layouts with one outputting characters in the normalization form C and one in the normalization form D.
+>
+> All keyboards in the CLDR repository will be in `NFC` form.  However, users and implementations may produce and consume other normalization forms.
+>
+> Tooling should verify that all possible keystrokes and gestures on the keyboard only produce the specified normalization form.
```

Comment:
Thinking about this a bit more, this is a little ambiguous. We can only really verify that individual strings are in the specified normalization form, not that all text produced is going to be in the correct form. For example:

```xml
<key id="a" to="a" />
<key id="mark" to="\u0300" />
```

Each of those outputs is by itself NFC. But press <kbd>a</kbd> and then <kbd>mark</kbd>, and that's of course no longer NFC. The implementation is supposed to handle normalization.

So the only thing I think it makes sense to validate here is:

1. all attribute values are normalized

Reply:
> Thinking about this a bit more, this is a little ambiguous. We can only really verify that individual strings are in the specified normalization form, not that all text produced is going to be in the correct form. For example:

That's a good point.

> The implementation is supposed to handle normalization.

That needs to be documented, then. 

> 1. all attribute values are normalized

well… but if we have:

```xml
<key id="c" to="a\u0300" />
```

couldn't the implementation *also* handle the normalization? just as it would for a+mark.

But I think it would be better/cleaner to say that all attribute values are in the specified form.

Reply:
> couldn't the implementation _also_ handle the normalization?

Yes it could. But it is messy as you say, and unnecessary, and leads to potentially confusing data for users.

Reply:
I believe we agreed that the declared normalization is only informative for UI purposes, i.e. we will not expect platforms to change the normalization forms used in the file. I would even argue that implementation should not change the normalization "on the fly" going against wishes of the layout author.

Reply:
@miloush, I have a vague memory of that discussion now. Here's what I found in the [minutes from 2020-08-25/26](https://docs.google.com/document/d/1ZtrHfr85vnAXxgD6pwyMUGGrOMRoHRxpIBGLArpaDdI/edit#heading=h.lh4or466oeyh):

> Spec proposal discussion: https://unicode-org.atlassian.net/browse/CLDR-13273
> * TR35 is worded to assume and prefer NFC
> * Tooling that converts LDML to platform implementations may prefer to convert NFC to NFD
> * TR 35 implementation note: When a sequence is consumed in the keyboard state machine that output is expected to be NFC for CLDR. Other implementations may choose to use LDML files with NFD form or non-conforming and mixed sequences
> * Add a metadata property to allow a keyboard author to specify the normalization style of the layout ( example:  normalization=NFC, normalization=NFD, normalization=mixed )
> * It is not required for a keyboard to anticipate and normalize the output from a sequence of LDML keystrokes 

You wrote:
> I would even argue that implementation should not change the normalization "on the fly" going against wishes of the layout author.

That matches the expectations of Richard Wordingham per the CLDR ticket above. What about normalization for input to transforms? If we don't do this, we force keyboards to deal with multiple forms of context strings, because input to transforms can be arbitrary data not generated by the keyboard, during edit sessions.

Also. I know that some apps already normalize input as they receive it, so that is outside of the control of the implementation.

Reply:
I would rather say that implementations will normalize to the specified format before processing the transform. It would add a lot of burden and inconsistencies on the transform to have to handle all possible forms. That also implies that the transform input strings can be validated as well, no need for  both `é` and `e´` in input.

Reply:
@miloush 

> …we will not expect platforms to change the normalization forms used in the file.

so if a keyboard is declared as `NFC` we wouldn't expect the platform to send it NFD as transform input, and we wouldn't expect the keyboard to output NFD. Correct?

>  I would even argue that implementation should not change the normalization "on the fly" going against wishes of the layout author

I think the implementation should stick to the form declared, yes.

in the example above, are you (Jan) suggesting is that authors of an NFC keyboard with:

```xml
<key id="a" to="a" />
<key id="mark" to="\u0300" />
```

… would need to then add a transform from `a\u0300` -> `à` otherwise the keyboard will output non-NFC? That seems like it would add more burden.

Reply:
> in the example above, are you (Jan) suggesting is that authors of an NFC keyboard with

Yeah I am suggesting that is the case. We could revisit the informativeness of the attribute, the feature I care about here is to have the option to not have the normalization being messed with, which could for example be indicated by "mixed" value, or perhaps if we do this, even better by value like "preserve".

We didn't really discuss the processing of inputs to transforms I think, but isn't that covered by "When a sequence is consumed in the keyboard state machine that output is expected to be NFC for CLDR."?